shell/common/shell.cc

```diff
@@ -2023,6 +2037,71 @@ bool Shell::OnServiceProtocolReloadAssetFonts(
   return true;
 }
 
+void Shell::AddView(int64_t view_id) {
+  TRACE_EVENT0("flutter", "Shell::AddView");
+  FML_DCHECK(is_set_up_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+  if (view_id == kFlutterImplicitViewId) {
+    FML_DLOG(WARNING)
+        << "Unexpected request to add the implicit view #"
+        << kFlutterImplicitViewId
+        << ". This view should never be added. This call is no-op.";
+    return;
+  }
+
+  task_runners_.GetUITaskRunner()->PostTask([engine = engine_->GetWeakPtr(),  //
+                                             view_id                          //
+  ] {
+    if (engine) {
+      engine->AddView(view_id);
+    }
+  });
+
+  task_runners_.GetRasterTaskRunner()->PostTask(
+      [rasterizer = rasterizer_->GetWeakPtr(),  //
+       view_id                                  //
+  ]() {
+        if (rasterizer) {
+          rasterizer->AddView(view_id);
+        }
+      });
+}
+
+void Shell::RemoveView(int64_t view_id) {
+  TRACE_EVENT0("flutter", "Shell::RemoveView");
+  FML_DCHECK(is_set_up_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+  if (view_id == kFlutterImplicitViewId) {
+    FML_DLOG(WARNING)
+        << "Unexpected request to remove the implicit view #"
+        << kFlutterImplicitViewId
+        << ". This view should never be removed. This call is no-op.";
+    return;
+  }
+
+  expected_frame_sizes_.erase(view_id);
+  fml::AutoResetWaitableEvent latch;
+  task_runners_.GetUITaskRunner()->PostTask([engine = engine_->GetWeakPtr(),  //
+                                             view_id                          //
+  ] {
+    if (engine) {
+      engine->RemoveView(view_id);
+    }
+  });
+
+  task_runners_.GetRasterTaskRunner()->PostTask(
+      [&latch,                                  //
+       rasterizer = rasterizer_->GetWeakPtr(),  //
+       view_id                                  //
+  ]() {
+        if (rasterizer) {
+          rasterizer->RemoveSurface(view_id);
+        }
+        latch.Signal();
+      });
+  latch.Wait();
```

Comment:
Do we need a second latch so that we can wait until the UI thread has removed its view?

Also, should the raster thread wait until the UI thread has removed its view? If not, could you add a quick comment explaining why this is safe?

Reply:
Duplicate issue as https://github.com/flutter/engine/pull/42991#discussion_r1280159459